src/hooks/useViewportRows.ts

```diff
@@ -0,0 +1,93 @@
+import { useMemo } from 'react';
+
+import { GroupRow, GroupByDictionary, Dictionary } from '../types';
+import { getVerticalRangeToRender } from '../utils';
+
+interface ViewportRowsArgs<R, SR> {
+  rawRows: readonly R[];
+  rowHeight: number;
+  clientHeight: number;
+  scrollTop: number;
+  groupBy: readonly string[];
+  rowGrouper?: (rows: readonly R[], columnKey: string) => Dictionary<R[]>;
+  expandedGroupIds?: Set<unknown>;
+}
+
+export function useViewportRows<R, SR>({
+  rawRows,
+  rowHeight,
+  clientHeight,
+  scrollTop,
+  groupBy,
+  rowGrouper,
+  expandedGroupIds
+}: ViewportRowsArgs<R, SR>) {
+  const [groupedRows, rowsCount] = useMemo(() => {
+    if (groupBy.length === 0 || !rowGrouper) return [undefined, rawRows.length];
+
+    const groupRows = (rows: readonly R[], [groupByKey, ...remainingGroupByKeys]: readonly string[], startRowIndex: number): [GroupByDictionary<R>, number] => {
+      let groupRowsCount = 0;
+      const groups: GroupByDictionary<R> = {};
+      for (const [key, childRows] of Object.entries(rowGrouper(rows, groupByKey))) {
+        // Recursively group each parent group
+        const [childGroups, childRowsCount] = remainingGroupByKeys.length === 0
+          ? [childRows, childRows.length]
+          : groupRows(childRows, remainingGroupByKeys, startRowIndex + groupRowsCount + 1); // 1 for parent row
+        groups[key] = { childRows, childGroups, startRowIndex: startRowIndex + groupRowsCount };
+        groupRowsCount += childRowsCount + 1; // 1 for parent row
+      }
+
+      return [groups, groupRowsCount];
+    };
+
+    return groupRows(rawRows, groupBy, 0);
+  }, [groupBy, rowGrouper, rawRows]);
+
+  const rows = useMemo(() => {
+    if (!groupedRows) return rawRows;
+
+    const expandGroup = (rows: GroupByDictionary<R> | R[], parentKey: string | undefined, level: number): Array<GroupRow<R> | R> => {
+      if (Array.isArray(rows)) return rows;
+      const flattenedRows: Array<R | GroupRow<R>> = [];
+      Object.keys(rows).forEach((key, posInSet, keys) => {
```

Comment:
Use native `for` loop?

Reply:
We need index for aria-attributes. I think it is okay to use `forEach` here

Reply:
Classic `for` would also support the index. However, we don't have any `break;`, so it's not critical. I am just thinking for whatever even slightly performance we can improve, perhaps we shall do that.

```typescript
const keys = Object.keys(rows);
for (let posInSet = 0; posInSet < keys.length; posInSet++) {
  ...
}
```